### PR TITLE
Add qwen2.5 vl

### PR DESCRIPTION
**Pull request by maximizemaxwell**
# What does this PR do?
Add support for Qwen-2.5-VL

## Part of Issue
#2814

### CODE REVIEW

**Comment by lucasjinreal** [comment]
Nice work, does the result verified? Also, would consider add quantization support maybe

**Comment by maximizemaxwell** [comment]
It's still on work though..

**Comment by lucasjinreal** [comment]
Oh. no

**Comment by akshayballal95** [comment]
Nice Work. This is definitely needed. But I see that the model uses conv3d which is not there yet in Candle. How are you planning to handle that?

**Comment by maximizemaxwell** [comment]
@akshayballal95  I’ve been struggling this for a while. While flattening the temporal dimension and using Conv2D (reshaping [B, T, C, H, W] to [B*T, C, H, W]) may be possible, it changes the computation and is incompatible with pretrained weights. Implementing native Conv3D support in Candle seems to be the proper solution.

**Comment by akshayballal95** [comment]
Well, we have been asking for it

https://github.com/huggingface/candle/issues/795#issuecomment-2419446822

**Comment by EricLBuehler** [comment]
@maximizemaxwell you can look at: https://github.com/EricLBuehler/mistral.rs/blob/4608202c128da44b84157573dbc8ff1a1146f64c/mistralrs-core/src/layers.rs#L1965-L2036

This is written under the assumption that the temporal patch size==`kernel_sizes[0]`==2, which is true for Qwen 2.5 VL.

**Comment by maximizemaxwell** [comment]
Gosh I didn’t realize there was already an issue — I just created a new one with the same content, thinking it didn’t exist.
For now, I think it’s best to focus on implementing conv3d and postpone support for Qwen-VL.
I’ll think through a concrete implementation plan for conv3d tomorrow and open a PR under the issue you shared.

**Comment by lucasjinreal** [comment]
We need someone to review these features as already implemented in the community. Please merge it into candle so that other people could bring more broad model support to candle!

File: candle-transformers/src/models/qwen2_5_vl.rs [hunk 1]
```diff
@@ -0,0 +1,918 @@
...
+
+        let residual = &x;
+        let x = self.post_attention_layernorm.forward(&x)?;
+        let x = self.mlp.forward(&x)?;
+        x + residual
+    }
+}
+
+#[derive(Debug, Clone)]
+struct VisionPatchEmbed {
```
Comment (EricLBuehler) [inline candle-transformers/src/models/qwen2_5_vl.rs:344]: PatchEmbed uses Conv2d instead of Conv3d
- Qwen2.5-VL requires Conv3d to handle temporal patches (video frames)
- Should be: Conv3dNoBias with kernel [temporal_patch_size, patch_size, patch_size]

File: candle-transformers/src/models/qwen2_5_vl.rs [hunk 2]
```diff
@@ -0,0 +1,918 @@
...
+    layers: Vec<VisionEncoderLayer>,
+    merger: Option<VisionPatchMerger>,
+}
+
+impl VisionModel {
+    pub fn new(config: &VisionConfig, vb: VarBuilder) -> Result<Self> {
+        let patch_embed = VisionPatchEmbed::new(config, vb.pp("patch_embed"))?;
+
+        let head_dim = config.hidden_size / config.num_attention_heads;
+        let rotary_emb = Arc::new(MultimodalRotaryEmbedding::new(
```
Comment (EricLBuehler) [inline candle-transformers/src/models/qwen2_5_vl.rs:427]: The vision model uses a standard rotary embedding layer, not `MultimodalRotaryEmbedding`.

File: candle-transformers/src/models/qwen2_5_vl.rs [hunk 3]
```diff
@@ -0,0 +1,918 @@
...
+            layers,
+            merger,
+        })
+    }
+
+    pub fn forward(&self, pixel_values: &Tensor) -> Result<Tensor> {
+        let mut x = self.patch_embed.forward(pixel_values)?;
+
+        for layer in &self.layers {
+            x = layer.forward(&x, &self.rotary_emb)?;
```
Comment (EricLBuehler) [inline candle-transformers/src/models/qwen2_5_vl.rs:464]: The vision model has an attention mask, with `fullatt_block_indexes` to decide fill/windowed.

File: candle-transformers/src/models/qwen2_5_vl.rs [hunk 4]
```diff
@@ -0,0 +1,918 @@
...
+    }
+
+    fn forward(&self, x: &Tensor) -> Result<Tensor> {
+        // x: [batch, channels, height, width]
+        self.proj.forward(x)?.flatten_from(2)?.transpose(1, 2) // [batch, num_patches, hidden_size]
+    }
+}
+
+#[derive(Debug, Clone)]
+struct VisionPatchMerger {
```
Comment (EricLBuehler) [inline candle-transformers/src/models/qwen2_5_vl.rs:373]: PatchMerger uses wrong normalization - it uses LayerNorm but should use RmsNorm

File: candle-transformers/src/models/qwen2_5_vl.rs [hunk 5]
```diff
@@ -0,0 +1,918 @@
...
+//! - 🤗 [Qwen2.5-VL Model](https://huggingface.co/Qwen/Qwen2.5-VL-7B-Instruct)
+//! - [Blog Post](https://qwenlm.github.io/blog/qwen2.5/)
+
+use crate::models::with_tracing::{linear, linear_no_bias, Linear};
+use candle::{DType, Device, IndexOp, Module, Result, Tensor, D};
+use candle_nn::{Conv2d, Conv2dConfig, Embedding, LayerNorm, VarBuilder};
+use std::sync::Arc;
+
+#[derive(Debug, Clone, PartialEq, serde::Deserialize)]
+pub struct VisionConfig {
```
Comment (EricLBuehler) [inline candle-transformers/src/models/qwen2_5_vl.rs:22]: This misses some critial configs (e.g. `window_size`, `fullatt_block_indexes`, ...)

File: candle-transformers/src/models/qwen2_5_vl.rs [hunk 6]
```diff
@@ -0,0 +1,918 @@
...
+
+        // Merge patches
+        let merge_size = 2; // Using default merge size
+        let x = x.reshape((b, h / merge_size, merge_size, w / merge_size, merge_size, c))?;
+        let x = x.permute((0, 1, 3, 2, 4, 5))?;
+        let x = x.flatten(3, 5)?;
+        let x = x.flatten(1, 2)?;
+
+        // Apply MLP
+        let x = self.mlp[0].forward(&x)?.gelu()?;
```
Comment (EricLBuehler) [inline candle-transformers/src/models/qwen2_5_vl.rs:409]: The activation should be loaded from the config and configurable (it is actually silu)

File: candle-transformers/src/models/qwen2_5_vl.rs [hunk 7]
```diff
@@ -0,0 +1,918 @@
...
+
+#[derive(Debug, Clone)]
+pub struct TextModel {
+    embed_tokens: Embedding,
+    layers: Vec<TextDecoderLayer>,
+    norm: RmsNorm,
+}
+
+impl TextModel {
+    pub fn new(config: &TextConfig, vb: VarBuilder) -> Result<Self> {
```
Comment (EricLBuehler) [inline candle-transformers/src/models/qwen2_5_vl.rs:684]: Text model weight paths missing `model` prefix.

**Review by EricLBuehler — CHANGES_REQUESTED** [review]
Hi @maximizemaxwell!

Thanks for the PR. Please see the review comments below. The current implementation seems to lack many components of the model. Please see:

https://github.com/EricLBuehler/mistral.rs/tree/master/mistralrs-core/src/vision_models/qwen2_5_vl
